### PR TITLE
discord server membership check

### PR DESCRIPTION
**Pull request by Chupalika**
how this works:
- added yet another tournament setting. Specify a Discord server ID to enforce membership when registering to the tournament
- the backend now logs in to the DIscord bot on startup. It will be used to check membership when registering
- if the user hasn't linked their Discord account, an error will be returned saying so
- if the user's Discord account was not found to be a member of the Discord server, a different error will be returned
- if some other error is encountered in the process, it'll return an error message saying to contact staff

### CODE REVIEW

File: server/api.ts [hunk 1]
```diff
@@ -180,6 +182,41 @@ router.postAsync("/register", ensure.loggedIn, async (req, res) => {
     );
     return res.status(400).send({ error: `Your country can't participate in this division` });
   }
+  
+  if (tourney.discordServerId) {
+    if (!req.user.discordId) {
+      logger.info(`${req.user.username} failed to register for ${req.body.tourney} (discord account not linked)`);
+      return res.status(400).send({ error: "You have not linked your Discord account." });
```
Comment (cory2067) [inline server/api.ts:189]: can we include some text to tell the user how they can link their discord account? (or ideally open user settings automatically)

File: client/src/components/pages/NewTourneyHome.js [hunk 2]
```diff
@@ -63,6 +63,7 @@ function NewTourneyHome({ tourney, user, setUser, setLoginAttention }) {
         lobbyMaxSignups: data.lobbyMaxSignups || 8,
         blacklist: (data.blacklist || []).toString(),
         requiredCountries: data.requiredCountries || [],
+        discordServerId: data.discordServerId || "",
```
Comment (cory2067) [inline client/src/components/pages/NewTourneyHome.js:66]: how do you find the discord server id/can this default to GTS server?

Comment (Chupalika) [reply client/src/components/pages/NewTourneyHome.js:66]: server id you can get by just right clicking a server and clicking "copy server id". I didn't really want to use a default ID cause I usually like to make things optional... but idk, i guess i could hard code in a default id, or maybe make it an input placeholder or display a note. up to you guys

File: server/api.ts (hunk 1 above)
Comment (Chupalika) [reply server/api.ts:189]: mmm I don't really like to open something automatically without context, so maybe an idea could be to show a popup message instead? something like, "Please link your Discord account in your user settings and then try again"

Comment (cory2067) [reply server/api.ts:189]: it doesn't have to open the settings modal, just a little more detail in that error message to tell the user where they should go to link their discord would probably be helpful

File: client/src/components/pages/NewTourneyHome.js (hunk 2 above)
Comment (cory2067) [reply client/src/components/pages/NewTourneyHome.js:66]: alright if it's that easy to get the id, then i guess we can leave it blank

**Review by cory2067 — APPROVED** [review]
looks good once test fixed (i.e. don't start discord bot if env is `test`)

**Pull request merged**